### ISSUE TITLE
Build failure with `--enable-g=all` and `pmix.h`

### ISSUE THREAD

**Issue by yfguo** [pmodels/mpich#3832]
@zhenggb72 has reported that when building with PMIX and `--enable-g=all`, the compilation fails with the following error:

```
/opt/ohpc/admin/pmix/include/pmix_common.h: In function 'pmix_value_destruct':
/opt/ohpc/admin/pmix/include/pmix_common.h:2074:34: warning: character constant too long for its type
             free((m)->data.string);
                                  ^
In file included from ./src/include/mpiimpl.h:183:0,
                 from src/mpi/attr/attr_delete.c:8:
./src/include/mpir_mem.h:92:27: warning: statement with no effect [-Wunused-value]
 #define free(a)           'Error use MPL_free'   :::
                           ^
```

After looking at the code of src/mpl/include/mpl_trmem.h, I think the problem is due to the `#define` of the `free()` function when memory tracing is enabled. What happens is `pmix.h` includes `pmix_common.h` which has macros that calls `free()`. In this case, these invocations will be incorrectly replaced by our `#define` and cause the build to fail.

**Comment by hzhou** [comment]
I think the idea of memory tracing need to first ensure all memory functions are called through the MPL layer, that is why it first make all *normal* memory function "error" first. 

Seems what we need here is to make sure `pmix_common.h` be included before the `mpl.h` or `mpir_mem.h` so it lies outside the memory tracing facility.

**Comment by raffenet** [comment]
Should we just remove the error defines? We've discussed this in the past, but this is the first concrete breakage we've run into in an external package.

**Comment by hzhou** [comment]
The error defines are necessary to ensure the effectiveness of memory tracing. A separate question is how useful is the memory tracing at all? Who has the experience using it? Does valgrind cover the use case?

**Comment by raffenet** [comment]
Memory tracing is enabled on `debug` builds in Jenkins. Tests will fail if memory is not properly freed, or if bad reads/writes are detected. Valgrind can cover these use cases, however our Valgrind Jenkins tests are too painfully slow to be used with any regularity.

**Comment by wesbland** [comment]
There’s also some additional functionality that I added a year or two ago that allows you to measure types of memory and "high water” marks for maximum usage.

**Comment by hzhou** [comment]
@wesbland "high water" marks sounds very interesting. How to use it?

**Comment by raffenet** [comment]
```export MPIR_CVAR_MEMDUMP=1```

**Comment by wesbland** [comment]
That’s why all of the memory allocation functions take a “class” now. So they can be bucketed for that dump.

**Comment by yfguo** [comment]
I agree that we can adjust the order of inclusion. Do we have a convention on the inclusion order? @raffenet I do seems to remember one.

**Comment by hzhou** [comment]
In my opinion, the ideal case (and the way it should be) is the order of headers should not matter. For the case that do -- I consider broken design -- just fix it on a case by case basis, and make a note of it.

**Comment by wesbland** [comment]
> In my opinion, the ideal case (and the way it should be) is the order of headers should not matter. For the case that do -- I consider broken design -- just fix it on a case by case basis, and make a note of it.

Unfortunately @hzhou , because of the design of CH4 and its heavy use of headers, inclusion order is very important.

**Comment by hzhou** [comment]
Right, those inline headers are not really the same as normal headers. Personally I would like to name them differently, such as `.inl`.

**Comment by yfguo** [comment]
The order is always important for being a C preprocessor. I am wondering if this would work.
```c
#define free(a) \
    #warning "Should use MPL_free()" \
    do { \
        free() \
    } while (0)
```

**Comment by wesbland** [comment]
> The order is always important for being a C preprocessor. I am wondering if this would work.
> 
> ```c
> #define free(a) \
>     #warning "Should use MPL_free()" \
>     do { \
>         free() \
>     } while (0)
> ```

Wouldn't this just change the problem from errors to warnings? The problem here is the inclusion ordering. Having it be an error forces us to fix it.

**Comment by hzhou** [comment]
I am making a pr for it.

**Comment by yfguo** [comment]
Nevermind, that code would never work.

**Comment by hzhou** [comment]
Okay, it is way messier than I expected. `pmix.h` contains both inline functions and macros. Inline functions are easy, they just need be included before our `mpl.h`. Macros are nasty, as they expanded at time of use. All the solutions I can think of are ugly and disgusting, so my final conclusion is this:

We'll not fix/work-around it in MPICH. We'll put a note here and leave this issue open, and point future unfortunate *users* to here.

Meanwhile, I will file a issue/complaint to pmix. It is their inconsiderate design and it is easy to fix on their end -- they could either make these macros inline functions or not to used naked memory functions -- i.e. use `pmix_free`, `pmix_malloc`, `pmix_calloc`, ... which can be `inlined` as `free`, `malloc`, or `calloc`.

**Comment by hzhou** [comment]
Fixed by PR#3855
